### ISSUE TITLE
Timesteps as computed by XBraid not exact

### ISSUE THREAD

**Issue by DanielDoehring** [XBraid/xbraid#79]
## Issue
I observed that the timesteps `dt` as computed by the[ C++ example](https://github.com/XBraid/xbraid/blob/c4cf91661c4291457f5cd6a5740eff4e0ff3db04/examples/ex-01-pp.cpp) are not exact, i.e., for a long enough runtime, some inaccuracies occur:

[# TimeSteps](https://github.com/XBraid/xbraid/blob/c4cf91661c4291457f5cd6a5740eff4e0ff3db04/examples/ex-01-pp.cpp#L313) = 10 (Example):
![image](https://user-images.githubusercontent.com/75639095/125951665-2865b5a0-92eb-40fb-8282-c69b0efcc808.png)

[# TimeSteps](https://github.com/XBraid/xbraid/blob/c4cf91661c4291457f5cd6a5740eff4e0ff3db04/examples/ex-01-pp.cpp#L313) = 10^4 (Altered):
![image](https://user-images.githubusercontent.com/75639095/125952063-e83a279a-4315-4d39-a239-e4cfa5f9517e.png)
For this number of timesteps, the second (index 1) level stays exact at 1. 

[# TimeSteps](https://github.com/XBraid/xbraid/blob/c4cf91661c4291457f5cd6a5740eff4e0ff3db04/examples/ex-01-pp.cpp#L313) = 10^5 (Altered):
For this number of timesteps, also the level 1 timestep becomes inaccurate:
![image](https://user-images.githubusercontent.com/75639095/125952788-b66a2279-9f76-45c2-a5a3-595d8fc8f0f8.png)
![image](https://user-images.githubusercontent.com/75639095/125952814-f1cf2cea-897a-4d60-b9ce-b2e3c7833940.png)

Interestingly, the deviations seem to average each other exactly out.

### Code changes:
Lines 64, 65:
```c++
#include <iostream> // For printing
#include <limits>   // std::numeric_limits
```

182 - 187:
```c++
int level;
pstatus.GetLevel(&level);

const double dt = tstop - tstart;
std::cout.precision(std::numeric_limits<double>::max_digits10); // Print entire double
std::cout << "STEP: dt on lvl: " << level << " is: " << dt << '\n';
```

And finally `ntime =` {10000, 100000}.
I attached the changed file as a txt, since github does not support attaching .cpp source files.
[ex-01-pp_Changed.txt](https://github.com/XBraid/xbraid/files/6830896/ex-01-pp_Changed.txt)

### XBraid (3.0.0)
The options are the same as in the example, I altered only the [number of timesteps](https://github.com/XBraid/xbraid/blob/c4cf91661c4291457f5cd6a5740eff4e0ff3db04/examples/ex-01-pp.cpp#L313).

### MPI Implementation & OS, C/++ compiler

- gcc/g++ 9.3.0 
- OpenMPI 3.1.3 on Linux CENTOS 7 
- OpenMPI 4.0.5 on Ubuntu 20.04 LTS

## Context

I encountered this issue while performing the "exactness" test (fourth debug step as listed in the XBraid User Manual) for an own project.
In our code, we pre-compute some matrices which depend on the time step. For setting up these matrices, we used exact, self-computed values. We then pre-factorize the matrices (e.g. LU) and re-use them in the main step, where the right-hand-side is computed with the `dt` as given by XBraid. 
Apparently, these deviations in `dt` lead eventually to a violation of the exactness property, i.e., the residual did not stay at 0.0 or machine accuracy.

**Comment by jbschroder** [comment]
Hi Daniel,

Good question!  I wouldn't call the dt values inexact, so much as
unexpected in your case.  When choosing a simple formula to compute our
time-grid, it's a little subtle.

We can't just let
dt = (t_final  -  t_initial) / (Ntime - 1)
because roundoff will often lead to dt being a bit off, and then that can
lead to real issues, like the Braid t_final not equaling the t_final you
input.

Instead, Braid computes each time-value (instead of dt) in line 109 of
https://github.com/XBraid/xbraid/blob/master/braid/braid.c
and because of this formula in 109, dt values can vary by one or two bits.
But, we do guarantee that our t_final is the same t_final as you input.
So, in the interest of keeping things simple, we've made this compromise
where dt can vary just a tiny bit.

If you really need to have a preset time-grid, you can use the
braid_SetTimeGrid(...);
option.  See example/ex-01-expanded.c.

I would say that if your exactness test passes for smaller number of
time-steps, that you're probably good to go.  But if you're nervous, then
use the SetTimeGrid(...) option.  You can't go wrong there.

Sorry for any confusion on this!  It is a subtle point.

Thanks,

Jacob

**Comment by DanielDoehring** [comment]
Hi Jacob,

thanks for the quick answer and the clarification!
I must somehow glossed over the `SetTimeGrid()` option - probably because I thought this option would be primarily designed for non-uniform temporal grids. 
I will try this option, thanks again for the suggestion - but can deviations for higher levels (as apparent for `ntime = 1e5`) also be avoided? 
The documentation / user manual mentions only the finest grid. 

Best,
Daniel

**Comment by jbschroder** [comment]
Hi Daniel,

Another good question!  The coarse grid time values are restricted fine
grid time values, so if you set the fine grid time values to what you want,
then they will carry down to the coarser levels.

Best,

Jacob

**Comment by DanielDoehring** [comment]
Alright, cool! 
Maybe it would provide useful adding a small note in the user manual mentioning this phenomenon?
I am not too sure where one could add this, but from a user perspective, the debugging instructions might be the right place.
Apart from that -at least form my side - this issue may then be closed.

**Comment by jbschroder** [comment]
Thanks Daniel!  I just updated the debugging section in the manual, as you
suggested.

Best,

Jacob

**Comment by jbschroder** [comment]
Addressed in documentation in commit 56c5de61b049ae927874d2c4ee89effef56dfdcf
